### ISSUE TITLE
simplify and improve repository-hdfs plugin

### ISSUE THREAD

**Issue by rmuir** [elastic/elasticsearch#15438]
As-is, the repository-hdfs plugin isn't going to work at all from what i see.

I'm just plotting what I think we need to do to fix it, before releasing it:

* Four plugin zip versions are created, none are really tested, none are working (except on windows). For example the `-lite.zip` has no hadoop jars and expects the users to drop hadoop jars in their ES_CLASSPATH. We don't allow this, this is not going to work at all.
* we need to drop the `-hadoop1` and `-lite` stuff and just one have one jar, we need to get the basics working, walk before we can run. from what i see, things aren't working at all.
* documentation recommends adding configuration files and similar to the `CLASSPATH` or ES `lib/` folder, we can't do things this way anymore! Its not going to work!
* security permissions are not correct: it only adds `jaas_nt`. This means that this plugin is only going to work on windows NT platforms: anything else = instant security exception.
* The use of HDFS should be made reliable (means hdfs2 only), since its a snapshot restore plugin. To me this means two things, fsync and atomic rename. I don't see any synchronization, and the rename it uses for committing does not look like the atomic one (rename2).
* no tests are failing!

**Comment by costin** [comment]
Pushed fixed [here](https://github.com/costin/elasticsearch/tree/hdfs-simplifications).

1. the plugin has been updated to be hadoop2 only (dropped `-hadoop1` and `lite`)
2. the plugin documentation does not mention `ES_CLASSPATH` or `lib/` - if this is mentioned anywhere, let me know
3. I have removed `hadoop-yarn-common` to prevent jar hell from occuring in the IDE
4. the MiniHDFS starts and stops a mini HDFS cluster. Its classpath is `testCompile` - I've tried creating a separate configuration however for some reason, it ends up with several dependencies with a different version that need to be specified in the resolution strategy and further more adds in jars that trigger jar hell (commons-beanutils vs commons-collection).`testCompile` has this already resolved so the configuration is smaller.
5. The permissions have been double-checked. I have tested this locally on windows only hence the `jaas_nt` permission. `nix is next
6. Updated the plugin code to use the `FileContext` class.

**Comment by costin** [comment]
I have updated the branch and rebased against master.

1. Since the branch is Hadoop2 only, I have replaced the aging `FileSystem` interface with `FileContext`. The underlying code proved to be separate and in many cases, the new features in `FileContext` were ignored by `FileSystem`. The new class in use should be more reliable in terms of fsync semantics (through `hsync).
2. the code explicitly writes blobs (creates new files) using `hsync`. Further more in case of a rename, it tries to `append` the file to trigger the `hsync` yet again.
3. The number of Hadoop2 jars required has been removed - turns out some were not really used.
4. Last but not least, the build system was improved a bit to start MiniHDFS before integration testing and to stop it afterwards. While it works when done manually, I couldn't get it so far to work properly in gradle:
a. Gradle starts the process in foreground which means the whole build is stopped. It needs to be wrapped to run in the background which is somewhat tricky because we don't want to compute the classpath ourselves but rather get it from gradle. ant.exec might help here instead of Gradle `JavaExec`
b. killing a process is tricky. I have the basic portable code in place (similar to @rjernst 's code) however I need to find a way to pass the pid info around and against, due to classpath issues (gradle vs the project) this might require some code duplication.

Basically 4 is the last item completely addressing the issue. @rjernst  do you have some free time to help? If not, I'll push further over the weekend to get Gradle to play nice.

**Comment by rmuir** [comment]
Thanks costin, ryan and I continued with your branch, he did some cleanup on dependencies, I am basically trying to help with the integration test and getting this working on unix systems.

The things resolved so far are:
1. Hadoop jars were added one by one, only the most minimal set with no transitive dependencies. this way we can get this thing working without jar hell etc and minimize the surface area. It is now up to ConnectException so I think its fairly close. 
2. Hadoop will not work at all on non-windows systems without full execute permission across the entire filesystem (https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/Shell.java#L381-L402). This is obviously unacceptable, I had to add a brutal hack to deliver a custom IOException from our security code (instead of SecurityException) when it tries to do this. This needs to ultimately be fixed in hadoop.
3. Hadoop needs `jaas_unix` loadLibrary permission or it fails with security exception (because only `jaas_nt` is added right now.

See in-progress changes here: https://github.com/elastic/elasticsearch/compare/master...rjernst:hdfs_without_transitive

One thing we have right now is a bug in our handling of slf4j which affects hadoop. We don't actually ship it in any ES distributions, yet the build treats it as `provided`, so its also not in any plugin that wants it. I hacked around this temporarily, but that needs to be sorted out.

I will try your minihdfs cluster now (manually) and try to get the integ  connecting / snapshotting / restoring. Ryan is working on hooking it into the integTest part of the build.

**Comment by rmuir** [comment]
@costin i converted core's rest tests (take snapshots and restores) and they all pass against the minihdfs now. As there is no longer jar hell, and we can do a real integration test, I will see if i can do further simplifications.

**Comment by costin** [comment]
Wow, thanks for the improvements - this must have been quite an effort. Tested it locally and it worked great (note that under Windows one needs `winuntils.exe` and the proper permissions for it but since Windows is really a second class citizen under Hadoop, I haven't added them to the policy file since as you mentioned, it's simply too dangerous (and unacceptable).
Glad to see the CL hack is no longer needed - makes things a lot easier.

I have updated (simplified) the docs [here](https://github.com/costin/elasticsearch/tree/hdfs2-only).
Is there anything else to push this into 2.2?

Thanks again for everything!

**Comment by rmuir** [comment]
Well there are a few things we should do for merging back to master: 

1. fully automate the new rest tests, so they can run like normal tests in jenkins on all platforms, etc. this means launching the hdfs minicluster as a proper test fixture. Right now they have hardcoded port numbers from the last time i ran it on my machine :) There is some new stuff around this here: #15561. I will look into this.

2. Ensure we only support `hdfs://`. We don't test anything else, and I know for a fact `file://` will not work, users should instead use the proper repo plugin which will be tested for that, and we should not expose hadoop VFS here in this plugin.  We should think about the configuration API here too, especially considering the warning in the docs about not putting paths in the URI. Instead of a URI at all, maybe we should just ask for the hdfs server. We can have an undocumented Setting that lets our hacky test filesystem still work for unit testing.

3. open bug report at hadoop so that their classes can be initialized under unix without blanket process execution permissions across the entire machine. Our hack is ok only as a temporary one.

4. fix the handling of `slf4j-api.jar` in our gradle build. the root cause is our gradle struggling on this `optional` dependency, really all are bogus and caused by the server having to perform double-duty as a client. currently the build is marking this as `provided` but it is in fact not provided, its excluded explicitly from all distributions. So it does not make its way into the HDFS plugin zip and then its missing. in this branch my hack just adds that jar to all distributions, I doubt that is what we really want to do.

as far as the docs go, I strongly dislike having docs that recommend the user modify plugin jars. What is the protocol compatibility issue and how often does it break in hadoop?

I have no clue about 2.2, I think we still have plenty of work to get this correct in master. Also 2.x would require additional maven/ant work to startup/shutdown hdfs during integration tests phase, etc.

**Comment by rmuir** [comment]
The miniHDFS is working now for integration tests. I forked your branch here: https://github.com/elastic/elasticsearch/compare/master...rmuir:hdfs2-only?expand=1

I will try to improve it now, but at least it is working and rest tests are passing with it.

**Comment by rmuir** [comment]
`gradle check` now works on both linux and os X and with reasonable speed. Windows is totally broken: unit tests do not work nor does the hdfs fixture work... I am working through those issues.

**Comment by rmuir** [comment]
On windows we can't start `hdfsFixture` correctly without having native libraries available. They aren't exactly packaged in any clean/efficient way and I don't think we want windows `dll` files flying around in our builds anyway.

So windows `integTest` skips the fixture and the snapshot/restore rest tests against it. Unit tests do work now though.

**Comment by rmuir** [comment]
`gradle hdfs` (the simple javaexec test) works under `test/fixtures/hdfs-fixture` if you setup hadoop native libraries. You must set HADOOP_HOME and put HADOOP_HOME/bin in the PATH. I used https://github.com/karthikj1/Hadoop-2.7.1-Windows-64-binaries/releases as a quick test in a VM. 

So this gives the possibility to test windows in jenkins, if some things are setup. It is optional and these things are still skipped if its not setup (only unit tests and simple "plugin is installed" rest test will run).

But the test fixture stuff is still totally broken on windows, it tries to do something crazy with a .bat file..., @rjernst needs to look.

**Comment by rmuir** [comment]
I reviewed the hdfs code and ensured it matches the same semantics as our FS repository and removed any kind of leniency/exists/etc, with the exception of https://github.com/elastic/elasticsearch/issues/15579 (we fail with FileAlreadyExistsException on purpose)

**Comment by rmuir** [comment]
I'm closing this. We are in fairly good shape in master... not perfect but the basics are there.
